### PR TITLE
fix: cache refreshToken promises

### LINKED ISSUE #338: Cache promise for refreshToken requests
We're getting reports of 1000s of connections to googleapis when a flood of new requests come in with an expired access token.  We should cache a single promise for refreshToken requests, and return that when the refresh is in transit.

### PR DESCRIPTION
**Pull request by JustinBeckwith**
Resolves #338

### CODE REVIEW

File: src/auth/oauth2client.ts [hunk 1]
```diff
@@ -503,6 +504,30 @@ export class OAuth2Client extends AuthClient {
    */
   protected async refreshToken(refreshToken?: string|
                                null): Promise<GetTokenResponse> {
+    if (refreshToken) {
```
Comment (ofrobots) [inline src/auth/oauth2client.ts]: nit: reverse this branch to flatten the meat of this method.

File: src/auth/refreshclient.ts [hunk 2]
```diff
@@ -65,9 +65,9 @@ export class UserRefreshClient extends OAuth2Client {
    * @param refreshToken An ignored refreshToken..
    * @param callback Optional callback.
    */
-  protected async refreshToken(refreshToken?: string|
-                               null): Promise<GetTokenResponse> {
-    return super.refreshToken(this._refreshToken);
+  protected async refreshTokenNoCache(refreshToken?: string|
+                                      null): Promise<GetTokenResponse> {
+    return super.refreshTokenNoCache(this._refreshToken!);
```
Comment (ofrobots) [inline src/auth/refreshclient.ts]: Why do you need the non-null operator? I don't quite follow how you can guarantee it is non-null anyway.

File: test/test.jwt.ts [hunk 3]
```diff
@@ -244,6 +244,54 @@ it('should refresh token if missing access token', (done) => {
...
+    nock('http://example.com').get('/').thrice().reply(200)
+  ];
+  const jwt = new JWT({
+    email: 'foo@serviceaccount.com',
+    keyFile: PEM_PATH,
+    scopes: ['http://bar', 'http://foo'],
+    subject: 'bar@subjectaccount.com'
+  });
+  jwt.credentials = {refresh_token: 'jwt-placeholder'};
+  jwt.credentials = {refresh_token: 'refresh-token-placeholder'};
```
Comment (ofrobots) [inline test/test.jwt.ts]: Both these lines needed?

File: src/auth/oauth2client.ts [hunk 4]
```diff
@@ -503,6 +504,30 @@ export class OAuth2Client extends AuthClient {
...
+        return this.refreshTokenPromises.get(refreshToken)!;
+      }
+
+      const p = this.refreshTokenNoCache(refreshToken)
+                    .then(r => {
+                      this.refreshTokenPromises.delete(refreshToken);
+                      return r;
+                    })
+                    .catch(e => {
+                      this.refreshTokenPromises.delete(refreshToken);
```
Comment (ofrobots) [inline src/auth/oauth2client.ts]: Is this catch path exercised by the tests?

Comment (JustinBeckwith) [reply src/auth/oauth2client.ts]: Nope.  Added one!

File: src/auth/refreshclient.ts (hunk 2 above)
Comment (JustinBeckwith) [reply src/auth/refreshclient.ts]: This wasn't needed

**Comment by JustinBeckwith** [comment]
Just tested out the fixed module with firebase, and I'm only seeing 5 connections over 1000 simultaneous calls.  Ready for another look @ofrobots :)

image

**Pull request merged**